src/network/socket_server.c

```diff
@@ -431,4 +384,280 @@ socket_server_config_t *new_socket_server_config(int num_addrs) {
+        return false;
+    }
+
+    msg->type = MESSAGE_BEGIN_ECDH;
+    msg->data[0] = 'o';
+    msg->data[1] = 'k';
+    msg->len = 2;
+
+    int rc = handle_send(data->srv->thl, data->conn->socket_number, msg);
+
+    free_message_t(msg);
+
+    if (rc == -1) {
+        return false;
+    }
+
+    return true;
+}
+
+/*!
+ * @brief handles receiving a hello protocol message from another peer
+ * @details is responsible for exchanging identification information with a peer
+ * @details and updating our peerstore with the appropriate information
+ */
+bool handle_hello_protocol(conn_handle_data_t *data, message_t *msg) {
+
+    // temporary struct to hold message data
+    cbor_encoded_data_t cbdata = {.data = msg->data, .len = msg->len};
+
+    // decode into the hello protocol message
+    message_hello_t *msg_hello = cbor_decode_hello_t(&cbdata);
+    if (msg_hello == NULL) {
+        data->srv->thl->log(data->srv->thl, 0,
+                            "failed to cbor decode message_hello_t",
+                            LOG_LEVELS_DEBUG);
+        return false;
+    }
+
+    // insert the peers information into our peerstore
+    bool ok = peerstore_insert_peer(data->srv->pstore, msg_hello->peer_id,
+                                    msg_hello->public_key, msg_hello->peer_id_len,
+                                    msg_hello->public_key_len);
+
+    free_message_hello_t(msg_hello);
+
+    if (ok == false) {
+        data->srv->thl->log(data->srv->thl, 0,
+                            "failed to insert peer into peerstore",
+                            LOG_LEVELS_DEBUG);
+        return ok;
+    } else {
+        data->srv->thl->log(data->srv->thl, 0,
+                            "successfully inserted peer into peerstore",
+                            LOG_LEVELS_DEBUG);
+    }
+    
+    // if this is MESSAGE_HELLO_FIN it means we dont need to continue the exchange
+    if (msg->type == MESSAGE_HELLO_FIN) {
+        return ok;
+    }
+
+    message_hello_t *send_msg_hello = new_server_message_hello_t(data->srv);
+    if (send_msg_hello == NULL) {
+        data->srv->thl->log(data->srv->thl, 0,
+                            "failed to create message_hello_t from server",
+                            LOG_LEVELS_DEBUG);
+        return false;
+    }
+
+    // send the MESSSAGE_HELLO_FIN message in the exchange
+    message_t *send_msg = message_hello_t_to_message_t(send_msg_hello, false);
+
+    free_message_hello_t(send_msg_hello);
+
+    if (send_msg == NULL) {
+        data->srv->thl->log(data->srv->thl, 0, "failed to get message_t",
+                            LOG_LEVELS_DEBUG);
+        return false;
+    }
+
+    // send the data to our peer
+    int rc = handle_send(data->srv->thl, data->conn->socket_number, send_msg);
+    if (rc == -1) {
+        data->srv->thl->log(data->srv->thl, 0, "failed to send message",
+                            LOG_LEVELS_DEBUG);
+        return false;
+    }
+
+    free_message_t(send_msg);
```

Comment:
move this to after `handle_send` otherwise is `handle_send` fails then this causes a memory leak